### DISCUSSION TITLE
Tarski/PDDL Parsing Issue - Equality

### DISCUSSION THREAD

**Discussion by beckydvn** [AI-Planning/macq#64] [General]
Upon examining the changes you suggested, I realized I left out a couple test cases - one of which is the 'equality' predicate provided by tarski. For example, when you supply a precondition/effect in the form of `(= ?genre1 ?genre2)` in tarski, you could get the grounded `=(jazz,rock)`.

Now the issue is that (as far as I know) it is impossible to tell the _types_ of the objects being compared during parsing, because equality is a 'default' predicate and can be used by any object. (For other user-provided predicates, I could determine the types of the objects easily because the predicate would always act on the same 'types' of objects - i.e. `on ?x ?y` checks if song y comes after song x and is obviously only applied to two objects of type `song`.) With equality though, I don't have that luxury. 

I figured it would be fine to just use a default `object` type for those objects, but the issue arises when those objects then need to be compared. When two CustomObjects are compared, the type of the objects as well as the name are checked for equality (`block` and `A` for example). The equality parsing issue results in the situation where `block A` and `object A` do not register as equal, even though `object A` _really is_ a block, there was just no way to tell.

As far as I can tell, there's a few ways to tackle this:

- Don't compare types when comparing objects. I don't like this solution because I feel like types are important and we shouldn't rely on names alone to compare them...
- Remove the need to compare objects at all. (As for now, I believe objects are only compared when adding preconditions/effects to an action to make sure that they match the action's parameters). The alternative would be to not check anything and simply add them to the list of the action's viable parameters whenever the user adds preconditions/effects. I don't see much issue with this one, unless you don't want to let the user change the parameters of an action.
- Some mystical way to extract types when given a 'default' predicate ;) That would be amazing, especially considering that it seems like there are other 'default' predicates which turns this into a larger issue.

**Comment by haz** [comment]
While I like to be mystical every chance I get, the solution to this one is no such thing. It's actually your first choice (the one you don't like). In PDDL, we assume that objects are distinct. You can't have two objects of the same name. If they need to be separate types, then you create a new type that inherits from both.

So, in this case, just toss the type and compare based on object name. Sorry it isn't so mystical!

**Reply by beckydvn** [reply]
Ok thanks - that is certainly the easiest solution so I'll take it 😄
